Tests/Unit/NetworkServer/ConcentratorDeduplicationTest.cs

```diff
@@ -95,30 +96,38 @@ public void When_Data_Message_Encountered_Should_Find_Duplicates_For_Different_D
             Assert.Equal(station1Eui, foundStation);
         }
 
-        public static TheoryData CreateKeyDataMessagesTheoryData
-            => TheoryDataFactory.From<object, ulong, ushort, ushort, string?>(
-                new (object, ulong, ushort, ushort, string?)[]
-                {
-                    (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(0), 0), 0, 0, 0, null),
-                    (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(0), 0), 0, 0, 0, "1"), // a non-relevant field should not influence the key
-                    (new ConcentratorDeduplication.DataMessageKey(new DevEui(0x1010101010101010UL), new Mic(0), 0), 0x1010101010101010UL, 0, 0, null),
-                    (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(1), 0), 0, 1, 0, null),
-                    (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(0), 1), 0, 0, 1, null)
-                }
-            );
+        public sealed class DataMessageKeyHolder
+        {
+            internal DataMessageKeyHolder(ConcentratorDeduplication.DataMessageKey dataMessageKey) => Value = dataMessageKey;
+            internal ConcentratorDeduplication.DataMessageKey Value { get; }
+        }
+
+        private static readonly (DevEui DevEui, Mic Mic, ushort FCnt, string? FieldNotUsedInKey)[] RawCreateKeyDataMessagesTheoryData = new (DevEui, Mic, ushort, string?)[]
+        {
+            (new DevEui(0), new Mic(0), 0,  null),
+            (new DevEui(0), new Mic(0), 0, "1"), // a non-relevant field should not influence the key
+            (new DevEui(0x1010101010101010UL), new Mic(0), 0, null),
+            (new DevEui(0), new Mic(1), 0, null),
+            (new DevEui(0), new Mic(0), 1, null)
+        };
+
+        public static TheoryData<DataMessageKeyHolder, DevEui, Mic, ushort, string?> CreateKeyDataMessagesTheoryData =>
+            TheoryDataFactory.From(from dataPoint in RawCreateKeyDataMessagesTheoryData
+                                   select (new DataMessageKeyHolder(new ConcentratorDeduplication.DataMessageKey(dataPoint.DevEui, dataPoint.Mic, dataPoint.FCnt)),
+                                           dataPoint.DevEui, dataPoint.Mic, dataPoint.FCnt, dataPoint.FieldNotUsedInKey));
```

Comment:
Not sure if all these indirections, additional type etc make the test easier to understand..why not just use the strong types for DevEui and Mic?

```suggestion
               public static TheoryData CreateKeyDataMessagesTheoryData
                    => TheoryDataFactory.From<ConcentratorDeduplication.DataMessageKey, DevEui, Mic, ushort, string?>(
                        new (ConcentratorDeduplication.DataMessageKey, DevEui, Mic, ushort, string?)[]
                        {
                            (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(0), 0), new DevEui(0), new Mic(0), 0, null),
                            (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(0), 0), new DevEui(0), new Mic(0), 0, "1"), // a non-relevant field should not influence the key
                            (new ConcentratorDeduplication.DataMessageKey(new DevEui(0x1010101010101010UL), new Mic(0), 0), new DevEui(0x1010101010101010UL), new Mic(0), 0, null),
                            (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(1), 0), new DevEui(0), new Mic(1), 0, null),
                            (new ConcentratorDeduplication.DataMessageKey(new DevEui(0), new Mic(0), 1), new DevEui(0), new Mic(0), 1, null)
                        }
                    );
```

Reply:
Added the clarification in the code with 67d7fa00cad9d5f9a86dec47326e92951cfb4feb

Reply:
ok, I see your point about the return type now (though I would say that we get already some type guarantees from the `TheoryDataFactory.From` method). Thanks for the explanation :)